packages/rpc/src/rpc-clusters.ts

```diff
@@ -2,28 +2,181 @@ import type { SolanaRpcApi, SolanaRpcApiDevnet, SolanaRpcApiMainnet, SolanaRpcAp
+ *
+ * This is useful in cases where you need to make assertions about what capabilities a RPC offers.
+ * For example, RPC methods like {@link requestAirdrop} are not available on mainnet. You can use
+ * the ability to assert on the type of RPC transport at compile time to prevent calling
+ * unimplemented methods or presuming the existence of unavailable capabilities.
+ */
 export type RpcTransportTestnet = RpcTransport & { '~cluster': 'testnet' };
+/**
+ * A {@link RpcTransport} that communicates with the mainnet cluster.
+ *
+ * Such transports are understood to communicate with a RPC server that services mainnet, and as
+ * such might only be accepted for use as the transport of a {@link RpcMainnet}.
+ *
+ * This is useful in cases where you need to make assertions about what capabilities a RPC offers.
+ * For example, RPC methods like {@link requestAirdrop} are not available on mainnet. You can use
+ * the ability to assert on the type of RPC transport at compile time to prevent calling
+ * unimplemented methods or presuming the existence of unavailable capabilities.
+ */
 export type RpcTransportMainnet = RpcTransport & { '~cluster': 'mainnet' };
+/**
+ * Given a {@link ClusterUrl}, this utility type will resolve to as specific a {@link RpcTransport}
+ * as possible.
+ *
+ * @example
+ * ```ts
+ * function createCustomTransport<TClusterUrl extends ClusterUrl>(
+ *     clusterUrl: TClusterUrl,
+ * ): RpcTransportFromClusterUrl<TClusterUrl> {
+ *     /* ... *\/
+ * }
+ *
+ * const transport = createCustomTransport(testnet('http://api.testnet.solana.com'));
+ * transport satisfies RpcTransportTestnet; // OK
+ * ```
+ */
 export type RpcTransportFromClusterUrl<TClusterUrl extends ClusterUrl> = TClusterUrl extends DevnetUrl
     ? RpcTransportDevnet
     : TClusterUrl extends TestnetUrl
       ? RpcTransportTestnet
       : TClusterUrl extends MainnetUrl
         ? RpcTransportMainnet
         : RpcTransport;
-
+/**
+ * A {@link Rpc} that supports the RPC methods available on the devnet cluster.
+ *
+ * This is useful in cases where you need to make assertions about the suitability of a RPC for a
+ * given purpose. For example, you might like to make it a type error to combine certain types with
+ * RPCs belonging to certain clusters, at compile time.
+ *
+ * @example
+ * ```ts
+ * async function getSpecialAccountInfo(
+ *     address: Address<'ReAL1111111111111111111111111111'>,
+ *     rpc: RpcMainnet,
+ * ): Promise<SpecialAccountInfo>;
+ * async function getSpecialAccountInfo(
+ *     address: Address<'TeST1111111111111111111111111111'>,
+ *     rpc: RpcDevnet | RpcTestnet,
+ * ): Promise<SpecialAccountInfo>;
+ * async function getSpecialAccountInfo(address: Address, rpc: Rpc): Promise<SpecialAccountInfo> {
+ *     /* ... *\/
+ * }
+ * const rpc = createSolanaRpc(devnet('https://api.devnet.solana.com'));
+ * await getSpecialAccountInfo(address('ReAL1111111111111111111111111111'), rpc); // ERROR
```

Comment:
Is this example correct? It looks like you're specifying an overload for `RpcDevnet` at 77-80. Should that just be `RpcTestnet`? 

(Just making the comment once, but the example is repeated below)

Reply:
I'm just saying that the `TeST111...` address is available on test clusters only. https://codesandbox.io/p/sandbox/ydjvcr

That said, the code is jacked for other reasons: https://app.graphite.dev/github/pr/anza-xyz/kit/240/Make-example-code-in-%60Rpc*%3C...%3E%60-types-runnable#